Update to 1.18

IMS212 opened the pull request:
No known issues,

IMS212 committed:
this is so incredibly hacky i hate everything about it

IMS212 committed:
Update to 21w40a

IMS212 committed:
fix build

Commit:
Update to 21w42a, add Autosave Indicator

Commit:
Add Unique to frame field

Commit:
Add support for custom fork of DataBreaker

Commit:
Remove references to the biome cache

Commit:
Update to 21w44a and Fabric API 0.42.1

Commit:
Update to Loom 0.10 release

PepperCode1 commented on src/main/java/me/jellysquid/mods/sodium/client/world/WorldSlice.java:
```diff
@@ -321,12 +313,7 @@ public Biome getBiomeForNoiseGen(int x, int y, int z) {
-
-        return this.biomeCaches[getLocalSectionIndex(relX >> 4, relY >> 4, relZ >> 4)]
-                .getBiome(this, x, relY >> 4, z);
+        return this.world.getBiome(new BlockPos(x, y, z));
```
Assuming `WorldSlice` instances are used on one thread at once, this should probably be replaced with a `BlockPos.Mutable` field being `.set` to the new position.

IMS212 replied:
What is better about this?

CalXee commented:
In my experience, loading chunks causes dark chunks quite often, and it's not just the spawn chunks (if you are referring to the thing I think you are referring to)
And PLEASE don't forget to add Chunk Builder (Initially known as Priority Updates) into the sodium GUI.

Commit:
Add Chunk Builder Mode and Simulation Distance options, fix autosave indicator translation

jellysquid3 commented:
>  And PLEASE don't forget to add Chunk Builder (Initially known as Priority Updates) and Simulation Distance into the sodium GUI.

Sodium completely replaces the chunk building code. We can't just use Minecraft's option here, it's implementation-specific and doesn't apply to Sodium in any way.

That said, in the experiments branch for Sodium 0.4 we have an "Always Defer Chunk Updates" option which gives something similar to what Mojang ended up adding later on in the MC 1.18 development cycle.

IMS212 commented:
I'll remove it, then. In the meantime, I've been looking into what's causing the black sections and I have no idea. All that I can tell is that 1.18 added a chunk update queue.

Commit:
Remove Chunk Builder Mode option

Commit:
Update to 1.18 pre2

Commit:
Update mixin compatibility level

Commit:
Update to 1.18-pre3

Commit:
Dark chunk fix attempt

IMS212 commented:
The dark chunks seem to have been fully fixed as of newest commit.

Commit:
Replace BlockPos with Mutable

Commit:
Create mutable only once

Commit:
Update to 1.18-pre5

Commit:
Merge branch '1.17.x/dev' into 1.18.x/dev

Commit:
Fix simulation distance setting

MESLewis commented on src/main/java/me/jellysquid/mods/sodium/client/gui/SodiumGameOptionPages.java:
```diff
@@ -38,6 +39,15 @@ public static OptionPage general() {
                         .setBinding((options, value) -> options.viewDistance = value, options -> options.viewDistance)
                         .setImpact(OptionImpact.HIGH)
                         .setFlags(OptionFlag.REQUIRES_RENDERER_RELOAD)
+                        .setFlags(OptionFlag.REQUIRES_RENDERER_RELOAD)
```
Duplicate line

MESLewis commented on src/main/java/me/jellysquid/mods/sodium/client/gui/SodiumGameOptionPages.java:
```diff
@@ -206,6 +216,13 @@ public static OptionPage quality() {
+                        .setName(new TranslatableText("options.autosaveIndicator"))
+                        .setTooltip(new TranslatableText("sodium.options.autosave_indicator.tooltip"))
+                        .setControl(TickBoxControl::new)
+                        .setBinding((opts, value) -> opts.showAutosaveIndicator /* Autosave indicator*/ = value, opts -> opts.showAutosaveIndicator)
```
Inline comment seems redundant

IMS212 replied:
That comment was added back when the autosave indicator field was unmapped, so yes, it is now redundant.

MESLewis commented on src/main/java/me/jellysquid/mods/sodium/mixin/features/chunk_rendering/MixinWorldRenderer.java:
```diff
@@ -11,10 +11,7 @@
-import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Overwrite;
-import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.*;
```
I think star imports are discouraged

MESLewis commented on src/main/resources/assets/sodium/lang/en_us.json:
```diff
@@ -28,6 +29,7 @@
   "sodium.options.entity_shadows.tooltip": "If enabled, basic shadows will be rendered beneath mobs and other entities.",
   "sodium.options.vignette.name": "Vignette",
   "sodium.options.vignette.tooltip": "If enabled, a vignette effect will be rendered on the player's view. This is very unlikely to make a difference to frame rates unless you are fill-rate limited.",
+  "sodium.options.autosave_indicator.tooltip": "If enabled, shows an \"Saving world\" indicator on the bottom right of your screen when auto-saving.",
```
I think 'a' instead of 'an' is better

Commit:
Fix some review issues

Removed inline comment, removed duplicate line, changed an to a, removed star import
